File: 2 Sorting/mergeSort.cpp
#include <iostream>
#include <vector>
using namespace std;

// Merge function to merge two sorted halves of arr[low..mid] and arr[mid+1..high]
void merge(vector<int>& arr, int low, int mid, int high) {
    // Temporary vector to store merged result
    vector<int> temp;

    int left = low;      // Pointer to the left half
    int right = mid + 1; // Pointer to the right half

    // Merge elements from both halves in sorted order
    while (left <= mid && right <= high) {
        if (arr[left] <= arr[right]) {
            temp.push_back(arr[left]);
            left++;
        } else {
            temp.push_back(arr[right]);
            right++;
        }
    }

    // If there are remaining elements in the left half
    while (left <= mid) {
        temp.push_back(arr[left]);
        left++;
    }

    // If there are remaining elements in the right half
    while (right <= high) {
        temp.push_back(arr[right]);
        right++;
    }

    // Copy merged elements back to original array
    for (int i = low; i <= high; i++) {
        arr[i] = temp[i - low]; // temp is 0-indexed; arr starts from 'low'
    }
}

// Recursive helper function for merge sort
void mS(vector<int>& arr, int low, int high) {
    if (low >= high) return; // Base case: single element is already sorted

    int mid = (low + high) / 2;

    // Sort left half
    mS(arr, low, mid);

    // Sort right half
    mS(arr, mid + 1, high);

    // Merge both sorted halves
    merge(arr, low, mid, high);
}

// Main merge sort function called from outside
void mergeSort(vector<int>& arr, int n) {
    mS(arr, 0, n - 1);
}

int main() {
    vector<int> arr = {5, 2, 9, 1, 5, 6};
    mergeSort(arr, arr.size());

    for (int num : arr) {
        cout << num << " ";
    }
    cout << endl;
    return 0;
}


/*


Merge Sort is O(n log n) time complexity.

It uses Divide and Conquer.

Space complexity is O(n) due to the temporary arrays.

It’s a stable sorting algorithm.



            [5, 2, 9, 1]
            /         \
         [5, 2]      [9, 1]
         /   \        /   \
       [5]  [2]     [9]  [1]

         ↓ merge     ↓ merge

       [2, 5]       [1, 9]

         ↓ merge

      [1, 2, 5, 9] ← Final sorted array


*/


/*


Initial Array:
[5, 2, 9, 1, 6, 3]

Dividing: [0, 5] -> [0, 2] and [3, 5]
Dividing: [0, 2] -> [0, 1] and [2, 2]
Dividing: [0, 1] -> [0, 0] and [1, 1]

Merging:
Left: [5]
Right: [2]
After Merge: [2, 5]

Merging:
Left: [2, 5]
Right: [9]
After Merge: [2, 5, 9]

Dividing: [3, 5] -> [3, 4] and [5, 5]
Dividing: [3, 4] -> [3, 3] and [4, 4]

Merging:
Left: [1]
Right: [6]
After Merge: [1, 6]

Merging:
Left: [1, 6]
Right: [3]
After Merge: [1, 3, 6]

Merging:
Left: [2, 5, 9]
Right: [1, 3, 6]
After Merge: [1, 2, 3, 5, 6, 9]

Sorted Array:
[1, 2, 3, 5, 6, 9]



*/

File: Kunal Kushwaha/10-Recursion/sorting/MergeSort.java
import java.util.Arrays;

/*
Merge Sort (top-down, recursive)

High-level steps:
1. Divide the array into two parts (left half, right half).
2. Recursively sort each half.
3. Merge the two sorted halves into a single sorted array.

Key points:
- At every "level" of recursion, all n elements are involved in merging.
- Depth of recursion tree is O(log n).
=> Time complexity: O(n log n)

Functional version:
- mergeSort(arr) returns a new sorted array.
- It does NOT sort the original array in-place.

In-place-like version:
- mergeSortInPlace(arr, start, end) uses the same arr, with index ranges.
- It still uses temporary array `mix` for merging, so auxiliary space is O(n),
  but it avoids creating new left/right subarrays via copyOfRange.
*/

public class MergeSort {
    public static void main(String[] args) {
        int[] arr = {5, 4, 3, 2, 1};

        // Functional style:
        // mergeSort(arr);             // This does nothing to 'arr' because we ignore the returned sorted array.
        // arr = mergeSort(arr);       // This would correctly reassign 'arr' to the sorted array.

        // “In-place” style (using index range [start, end)):
        mergeSortInPlace(arr, 0, arr.length);

        System.out.println(Arrays.toString(arr)); // Expect: [1, 2, 3, 4, 5]
    }

    // Functional merge sort: returns a NEW sorted array.
    static int[] mergeSort(int[] arr) {
        // Base case: an array of length 1 is already sorted.
        if (arr.length == 1) {
            return arr;
        }

        int mid = arr.length / 2;

        // Recursively sort the left half: arr[0..mid-1]
        int[] left = mergeSort(Arrays.copyOfRange(arr, 0, mid));

        // Recursively sort the right half: arr[mid..end-1]
        int[] right = mergeSort(Arrays.copyOfRange(arr, mid, arr.length));

        // Merge two sorted halves into one sorted array.
        return merge(left, right);
    }

    // Merge two sorted arrays 'first' and 'second' into a single sorted array.
    private static int[] merge(int[] first, int[] second) {
        int[] mix = new int[first.length + second.length];

        int i = 0; // pointer in first
        int j = 0; // pointer in second
        int k = 0; // pointer in mix

        // While both arrays have elements, pick the smaller one.
        while (i < first.length && j < second.length) {
            if (first[i] < second[j]) {
                mix[k] = first[i];
                i++;
            } else {
                mix[k] = second[j];
                j++;
            }
            k++;
        }

        // Copy remaining elements from 'first', if any.
        while (i < first.length) {
            mix[k] = first[i];
            i++;
            k++;
        }

        // Copy remaining elements from 'second', if any.
        while (j < second.length) {
            mix[k] = second[j];
            j++;
            k++;
        }

        return mix;
    }

    // In-place style merge sort using the same array 'arr' and index range [start, end)
    static void mergeSortInPlace(int[] arr, int start, int end) {
        // Base case: segment of length 1 is already sorted.
        if (end - start == 1) {
            return;
        }

        int mid = (start + end) / 2;

        // Sort the left half: [start, mid)
        mergeSortInPlace(arr, start, mid);

        // Sort the right half: [mid, end)
        mergeSortInPlace(arr, mid, end);

        // Merge the two sorted halves back into arr[start..end-1]
        mergeInPlace(arr, start, mid, end);
    }

    // Merge two sorted subarrays of arr:
    // left:  [start, mid)
    // right: [mid, end)
    // into a temporary array 'mix', then copy back to arr[start..end-1].
    private static void mergeInPlace(int[] arr, int start, int mid, int end) {
        int[] mix = new int[end - start];

        int i = start; // pointer in left half
        int j = mid;   // pointer in right half
        int k = 0;     // pointer in mix

        // Merge two sorted parts into mix[]
        while (i < mid && j < end) {
            if (arr[i] < arr[j]) {
                mix[k] = arr[i];
                i++;
            } else {
                mix[k] = arr[j];
                j++;
            }
            k++;
        }

        // Copy remaining elements from left half
        while (i < mid) {
            mix[k] = arr[i];
            i++;
            k++;
        }

        // Copy remaining elements from right half
        while (j < end) {
            mix[k] = arr[j];
            j++;
            k++;
        }

        // Copy the merged result back into the original array segment
        for (int l = 0; l < mix.length; l++) {
            arr[start + l] = mix[l];
        }
    }
}
